### PR TITLE
feat(xrpc): register variable rpc_time

### PR DESCRIPTION
**Pull request by tzssangglass**
### Description

Fixes # (issue)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Review by spacewander — COMMENTED** [review]
We can apply the patch:
```
index df6e430c..5bf3daa5 100644
--- apisix/core/ctx.lua
+++ apisix/core/ctx.lua
@@ -278,11 +278,12 @@ do
             else
                 local getter = apisix_var_names[key]
                 if getter then
+                    local ctx = t._ctx
                     if getter == true then
-                        val = ngx.ctx.api_ctx and ngx.ctx.api_ctx[key]
+                        val = ctx and ctx[key]
                     else
                         -- the getter is registered by ctx.register_var
-                        val = getter(ngx.ctx.api_ctx)
+                        val = getter(ctx)
                     end

                 else
@@ -341,6 +342,7 @@ function _M.set_vars_meta(ctx)
     end

     var._request = get_request()
+    var._ctx = ctx
     setmetatable(var, mt)
     ctx.var = var
 end
diff --git t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
index 28fa4abc..3ea0c7ea 100644
--- t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
+++ t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
@@ -30,12 +30,7 @@ local str_byte = string.byte


 core.ctx.register_var("rpc_len", function(ctx)
-    local curr_ctx = ctx.xrpc_session and ctx.xrpc_session._curr_ctx
-    if not curr_ctx then
-        core.log.warn("can't find current context")
-        return nil
-    end
-    return curr_ctx.len
+    return ctx.len
 end)

 local _M = {}
```

**Review by spacewander — COMMENTED** [review]
We can apply the patch:
```
index df6e430c..5bf3daa5 100644
--- apisix/core/ctx.lua
+++ apisix/core/ctx.lua
@@ -278,11 +278,12 @@ do
             else
                 local getter = apisix_var_names[key]
                 if getter then
+                    local ctx = t._ctx
                     if getter == true then
-                        val = ngx.ctx.api_ctx and ngx.ctx.api_ctx[key]
+                        val = ctx and ctx[key]
                     else
                         -- the getter is registered by ctx.register_var
-                        val = getter(ngx.ctx.api_ctx)
+                        val = getter(ctx)
                     end

                 else
@@ -341,6 +342,7 @@ function _M.set_vars_meta(ctx)
     end

     var._request = get_request()
+    var._ctx = ctx
     setmetatable(var, mt)
     ctx.var = var
 end
diff --git t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
index 28fa4abc..3ea0c7ea 100644
--- t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
+++ t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
@@ -30,12 +30,7 @@ local str_byte = string.byte


 core.ctx.register_var("rpc_len", function(ctx)
-    local curr_ctx = ctx.xrpc_session and ctx.xrpc_session._curr_ctx
-    if not curr_ctx then
-        core.log.warn("can't find current context")
-        return nil
-    end
-    return curr_ctx.len
+    return ctx.len
 end)

 local _M = {}
```

**Comment by tzssangglass** [comment]
> We can apply the patch:
> 
> ```
> index df6e430c..5bf3daa5 100644
> --- apisix/core/ctx.lua
> +++ apisix/core/ctx.lua
> @@ -278,11 +278,12 @@ do
>              else
>                  local getter = apisix_var_names[key]
>                  if getter then
> +                    local ctx = t._ctx
>                      if getter == true then
> -                        val = ngx.ctx.api_ctx and ngx.ctx.api_ctx[key]
> +                        val = ctx and ctx[key]
>                      else
>                          -- the getter is registered by ctx.register_var
> -                        val = getter(ngx.ctx.api_ctx)
> +                        val = getter(ctx)
>                      end
> 
>                  else
> @@ -341,6 +342,7 @@ function _M.set_vars_meta(ctx)
>      end
> 
>      var._request = get_request()
> +    var._ctx = ctx
>      setmetatable(var, mt)
>      ctx.var = var
>  end
> diff --git t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
> index 28fa4abc..3ea0c7ea 100644
> --- t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
> +++ t/xrpc/apisix/stream/xrpc/protocols/pingpong/init.lua
> @@ -30,12 +30,7 @@ local str_byte = string.byte
> 
> 
>  core.ctx.register_var("rpc_len", function(ctx)
> -    local curr_ctx = ctx.xrpc_session and ctx.xrpc_session._curr_ctx
> -    if not curr_ctx then
> -        core.log.warn("can't find current context")
> -        return nil
> -    end
> -    return curr_ctx.len
> +    return ctx.len
>  end)
> 
>  local _M = {}
> ```

update, thanks for pointing out a better way.

File: apisix/stream/xrpc/runner.lua [hunk 1]
```diff
@@ -91,7 +96,7 @@ end
 
 local function finialize_req(protocol, session, ctx)
     ctx._rpc_end_time = ngx_now()
-
+    session._curr_ctx = ctx
```
Comment (spacewander) [inline apisix/stream/xrpc/runner.lua:99]: We don't need it now?

Comment (tzssangglass) [reply apisix/stream/xrpc/runner.lua]: My mistake, I forgot to delete it. Deleted.

**Review by spacewander — APPROVED** [review]

**Review by bzp2010 — APPROVED** [review]

**Pull request merged**